refactor: Move calculation logic out from `CheckSequenceLocksAtTip()`

hebasto opened the pull request:
This PR is follow up for bitcoin/bitcoin#22677 and bitcoin/bitcoin#23683.

On master (013daed9acca1b723f599d63ab36b9c2a5c60e5f) it is not obvious that `CheckSequenceLocksAtTip()` function can modify its `LockPoints* lp` parameter which leads to https://github.com/bitcoin/bitcoin/pull/22677#discussion_r762040101.

This PR:
- separates the lockpoint calculate logic from `CheckSequenceLocksAtTip()` function into a new `CalculateLockPointsAtTip()` one
- cleans up the `CheckSequenceLocksAtTip()` function interface
- makes code easier to reason about (hopefully)

hebasto commented:
Rebased 154307c10b0d3c0e7dfbe463de7d040b2eb97c74 -> c4476a5ea787e3722d8c1e5fed5f0c6e2a554327 ([pr23897.01](https://github.com/hebasto/bitcoin/commits/pr23897.01) -> [pr23897.02](https://github.com/hebasto/bitcoin/commits/pr23897.02)) on top of the merged #23683.

hebasto commented:
Rebased c4476a5ea787e3722d8c1e5fed5f0c6e2a554327 -> 220475bdbca0598c6f5158e870b89ff9503a7258 ([pr23897.02](https://github.com/hebasto/bitcoin/commits/pr23897.02) -> [pr23897.03](https://github.com/hebasto/bitcoin/commits/pr23897.03)) on top of the merged #23976.

glozow commented on src/validation.cpp:
```diff
@@ -211,14 +211,75 @@ bool CheckFinalTx(const CBlockIndex* active_chain_tip, const CTransaction &tx, i
+    const CCoinsView& coins_view,
+    const CTransaction& tx,
+    int flags,
+    LockPoints* lp)
```
It would be nice to remove the raw pointers while we're changing the interface anyway. I think a better interface is to have `CalculateLockPoints()` return a `std::optional<LockPoints>` which encodes the same information as the bool.

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/23897#issuecomment-1102831191).

glozow commented on src/validation.cpp:
```diff
@@ -211,14 +211,75 @@ bool CheckFinalTx(const CBlockIndex* active_chain_tip, const CTransaction &tx, i
+        const CTxIn& txin = tx.vin[txinIndex];
+        Coin coin;
+        if (!coins_view.GetCoin(txin.prevout, coin)) {
+            return error("%s: Missing input", __func__);
```
Can be replaced with a `LogPrintf` and `return std::nullopt`. Also seems like this log could be improved by adding the transaction id and/or prevout that is missing.

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/23897#issuecomment-1102831191).

glozow commented on src/validation.cpp:
```diff
@@ -211,14 +211,75 @@ bool CheckFinalTx(const CBlockIndex* active_chain_tip, const CTransaction &tx, i
-                        int flags,
-                        LockPoints* lp,
-                        bool useExistingLockPoints)
+                        const LockPoints* lp)
```
Same here, would be nice to remove the raw pointer and make this parameter a `const LockPoints& lp`. It doesn't make sense to allow a nullptr `lp` to be passed in anymore, since we can't check sequence locks on empty lockpoints anyway.

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/23897#issuecomment-1102831191).

glozow commented on src/validation.cpp:
```diff
@@ -365,18 +377,13 @@ void CChainState::MaybeUpdateMempoolForReorg(
-        } else if (!validLP) {
-            // If CheckSequenceLocks succeeded, it also updated the LockPoints.
-            // Now update the mempool entry lockpoints as well.
+        if (!validLP) {
```
Note that we do have a bit of a behavior change here: we now update the mempool entry LockPoints even if we're going to remove the entry later. I'm not sure if this is desirable.

I suggest adding documentation as well, e.g. "If the cached LockPoints are now invalid, recalculate them and update the mempool entry before calling `CheckSequenceLocks()`.

glozow reviewed (commented):
Concept ACK, I agree that calculating lockpoints and checking them against the current chain should be separate methods and it was definitely unexpected that `CheckSequenceLocks()` was mutating the lockpoints passed in. There is a slight behavior change to `MaybeUpdateMempoolForReorg()` here.

hebasto commented:
Updated 220475bdbca0598c6f5158e870b89ff9503a7258 -> aa8f68b2352dd37cae59867276b38127d7041685 ([pr23897.03](https://github.com/hebasto/bitcoin/commits/pr23897.03) -> [pr23897.04](https://github.com/hebasto/bitcoin/commits/pr23897.04)):
- rebased
- addressed @glozow's comments


Also the PR description has been updated.

hebasto commented:
Rebased aa8f68b2352dd37cae59867276b38127d7041685 -> 5a6ce8cca31b3105cb8e96cd3e5e6c0da0fb3479 ([pr23897.04](https://github.com/hebasto/bitcoin/commits/pr23897.04) -> [pr23897.05](https://github.com/hebasto/bitcoin/commits/pr23897.05)) due to conflict with #24804.

glozow commented on src/validation.h:
```diff
@@ -289,6 +289,24 @@ PackageMempoolAcceptResult ProcessNewPackage(CChainState& active_chainstate, CTx
+ *                              responsible for holding the appropriate locks to ensure that
+ *                              calls to GetCoin() return correct coins.
+ *
+ * @returns The resulting height and time calculated and the hash of the block needed for calculation.
```
in fc31aa12b9599f3a23c22fd65b3eaa73dbb05d81:

Should document that this returns std::nullopt if there is an error, i.e., if inputs cannot be found in the `coins_view`

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/23897#issuecomment-1191310762).

glozow commented on src/validation.cpp:
```diff
@@ -185,11 +185,67 @@ bool CheckFinalTxAtTip(const CBlockIndex* active_chain_tip, const CTransaction&
+                                                   const CCoinsView& coins_view,
+                                                   const CTransaction& tx)
+{
+    assert(tip);
```
To me, this suggests `tip` should be a reference instead of a pointer?

hebasto replied:
`CChain::Tip()` is been passed as an argument, and it could be a `nullptr`.

glozow replied:
Sorry I should have said the *parameter* `tip`, as in, if a function expects a non-nullptr tip, then perhaps it should take a reference instead of a pointer. But I think I get what you mean, that `CChain::Tip()` returns a pointer and that's passed around pretty frequently in the codebase.

glozow commented on src/validation.h:
```diff
@@ -300,17 +300,26 @@ bool CheckFinalTxAtTip(const CBlockIndex* active_chain_tip, const CTransaction&
+ * Check if transaction will be BIP68 final in the next block to be created on top of tip.
+ * @param[in]   tip             Chain tip to check tx sequence locks against. For example,
+ *                              the tip of the current active chain.
+ * @param[in]   lp              Relative locktime constraints (BIP68).
```
in 5a6ce8cca31b3105cb8e96cd3e5e6c0da0fb3479

```suggestion
 * @param[in]   lp              LockPoints containing the height and time at which this transaction is final.
```

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/23897#issuecomment-1191310762).

glozow reviewed (commented):
code review ACK 5a6ce8cca31b3105cb8e96cd3e5e6c0da0fb3479

This looks correct to me, a few suggestions to make things clearer. Probably at least 1 other person should take a look as well.

hebasto commented:
Updated 5a6ce8cca31b3105cb8e96cd3e5e6c0da0fb3479 -> 803869954e0e3bc32279dac71cf40a0614b79566 ([pr23897.05](https://github.com/hebasto/bitcoin/commits/pr23897.05) -> [pr23897.06](https://github.com/hebasto/bitcoin/commits/pr23897.06), [diff](https://github.com/hebasto/bitcoin/compare/pr23897.05..pr23897.06)):

- addressed @glozow's comments

maflcko reviewed (commented):
Not sure.

* The functions are only ever called together, so it seems odd to pull them apart? No objection if you want to move-only some stuff out of a function and then call the newly created function in place.
* I think the previous changes/docs made it clear that lp may be modified? Otherwise, an inline comment seems fine as well?

glozow commented:
reACK 803869954e0e3bc32279dac71cf40a0614b79566

Only changes are doc updates, for https://github.com/bitcoin/bitcoin/pull/23897#discussion_r926451573 and https://github.com/bitcoin/bitcoin/pull/23897#discussion_r926486212

glozow commented:
> No objection if you want to move-only some stuff out of a function and then call the newly created function in place.

fwiw this is >50% move-only.

hebasto commented:
Rebased 803869954e0e3bc32279dac71cf40a0614b79566 -> 32e3fc173410641569bc4df805d5431cf0013d74 ([pr23897.06](https://github.com/hebasto/bitcoin/commits/pr23897.06) -> [pr23897.07](https://github.com/hebasto/bitcoin/commits/pr23897.07)) due to the conflict with #25677.

hebasto commented:
Rebased 32e3fc173410641569bc4df805d5431cf0013d74 -> bebaddb0767b9c52507e8d50a18510b592c14778 ([pr23897.07](https://github.com/hebasto/bitcoin/commits/pr23897.07) -> [pr23897.08](https://github.com/hebasto/bitcoin/commits/pr23897.08)) due to the conflict with #25073.

aureleoules reviewed (approved):
ACK bebaddb0767b9c52507e8d50a18510b592c14778
The code is now much clearer and easier to understand and the behavior doesn't seem to have changed.

hebasto commented:
Updated bebaddb0767b9c52507e8d50a18510b592c14778 -> ed2d714bd1130291a2f781e644f787d421cdf26e ([pr23897.08](https://github.com/hebasto/bitcoin/commits/pr23897.08) -> [pr23897.09](https://github.com/hebasto/bitcoin/commits/pr23897.09), [diff](https://github.com/hebasto/bitcoin/compare/pr23897.08..pr23897.09)):

- addressed @aureleoules's comments

hebasto commented on src/validation.cpp:
```diff
@@ -176,11 +176,67 @@ bool CheckFinalTxAtTip(const CBlockIndex& active_chain_tip, const CTransaction&
+        }
+    }
+
+    const std::pair<int, int64_t> lock_pair{CalculateSequenceLocks(tx, STANDARD_LOCKTIME_VERIFY_FLAGS, prevheights, index)};
```
Thanks!.[Updated](https://github.com/bitcoin/bitcoin/pull/23897#issuecomment-1278917666).

hebasto commented on src/validation.cpp:
```diff
@@ -176,6 +176,65 @@ bool CheckFinalTxAtTip(const CBlockIndex& active_chain_tip, const CTransaction&
+        }
+        if (coin.nHeight == MEMPOOL_HEIGHT) {
+            // Assume all mempool transaction confirm in the next block
+            prevheights[txinIndex] = tip->nHeight + 1;
```
Thanks!.[Updated](https://github.com/bitcoin/bitcoin/pull/23897#issuecomment-1278917666).

aureleoules reviewed (approved):
reACK ed2d714bd1130291a2f781e644f787d421cdf26e

stickies-v commented on src/validation.cpp:
```diff
@@ -176,11 +176,67 @@ bool CheckFinalTxAtTip(const CBlockIndex& active_chain_tip, const CTransaction&
+                                                   const CCoinsView& coins_view,
+                                                   const CTransaction& tx)
+{
+    assert(tip);
```
Would it make sense to pass a `CBlockIndex&` instead and avoid checking for nullptr here? If this fn requires a non-nullptr `CBlockIndex`, I think the fn signature should reflect that and the callsite to take responsibility.

hebasto replied:
Going to leave it as this to keep this PR in more reviewable state.

stickies-v commented on src/validation.h:
```diff
@@ -271,17 +271,28 @@ bool CheckFinalTxAtTip(const CBlockIndex& active_chain_tip, const CTransaction&
+ */
+std::optional<LockPoints> CalculateLockPointsAtTip(CBlockIndex* tip,
+                                                   const CCoinsView& coins_view,
+                                                   const CTransaction& tx);
```
nit: missing doxygen for `tx` param

stickies-v reviewed (commented):
Concept ACK. Hooray for clearer function intents and effects!

stickies-v reviewed (commented):
Approach ACK ed2d714bd

I think this is already a great step in the right direction, but imo there are a bunch more refactoring maintainability improvements in the touched functions that we can make. Quite a few of my comments are suggestions in that direction - I realize some may be out of scope. Even though I'd like to see them incorporated, I'm not very fussed about necessarily doing them in this PR - I'm happy working on a follow-up too.

stickies-v commented on src/validation.h:
```diff
@@ -271,17 +271,28 @@ bool CheckFinalTxAtTip(const CBlockIndex& active_chain_tip, const CTransaction&
+ *                              the tip of the current active chain.
+ * @param[in]   lp              LockPoints containing the height and time at which this
+ *                              transaction is final.
  * Simulates calling SequenceLocks() with data from the tip passed in.
  * Optionally stores in LockPoints the resulting height and time calculated and the hash
  * of the block needed for calculation or skips the calculation and uses the LockPoints
  * passed in for evaluation.
  * The LockPoints should not be considered valid if CheckSequenceLocksAtTip returns false.
```
@hebasto I think we don't modify LockPoints anymore?

```
 * Optionally stores in LockPoints the resulting height and time calculated and the hash
 * of the block needed for calculation or skips the calculation and uses the LockPoints
 * passed in for evaluation.
```

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/23897#issuecomment-1330960296).

hebasto commented:
Updated ed2d714bd1130291a2f781e644f787d421cdf26e -> 1636bc465fd96b9561de3dd21bed7f989f409de1 ([pr23897.09](https://github.com/hebasto/bitcoin/commits/pr23897.09) -> [pr23897.10](https://github.com/hebasto/bitcoin/commits/pr23897.10), [diff](https://github.com/hebasto/bitcoin/compare/pr23897.09..pr23897.10)):

- addressed some of @stickies-v's comments

hebasto commented on src/validation.cpp:
```diff
@@ -176,11 +176,67 @@ bool CheckFinalTxAtTip(const CBlockIndex& active_chain_tip, const CTransaction&
+        if (height != index.nHeight) {
+            max_input_height = std::max(max_input_height, height);
+        }
+    }
```
Going to leave it as this to keep this PR in more reviewable state.

hebasto commented on src/validation.cpp:
```diff
@@ -827,7 +832,7 @@ bool MemPoolAccept::PreChecks(ATMPArgs& args, Workspace& ws)
 
     entry.reset(new CTxMemPoolEntry(ptx, ws.m_base_fees, nAcceptTime, m_active_chainstate.m_chain.Height(),
-            fSpendsCoinbase, nSigOpsCost, lp));
+            fSpendsCoinbase, nSigOpsCost, *lock_points));
```
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/23897#issuecomment-1330960296).

hebasto commented on src/validation.cpp:
```diff
@@ -334,20 +336,23 @@ void Chainstate::MaybeUpdateMempoolForReorg(
 
         // The transaction must be final.
```
Going to leave it as this to keep this PR in more reviewable state.

hebasto commented on src/validation.cpp:
```diff
@@ -176,11 +176,67 @@ bool CheckFinalTxAtTip(const CBlockIndex& active_chain_tip, const CTransaction&
+        const CTxIn& txin = tx.vin[txinIndex];
+        Coin coin;
+        if (!coins_view.GetCoin(txin.prevout, coin)) {
+            LogPrintf("ERROR: %s: Missing input %d in transaction \'%s\'\n", __func__, txinIndex, tx.GetHash().GetHex());
```
This change could be considered as a behavior change, so leaving it for now.

hebasto commented on src/validation.cpp:
```diff
@@ -176,11 +176,67 @@ bool CheckFinalTxAtTip(const CBlockIndex& active_chain_tip, const CTransaction&
-                        const CTransaction& tx,
-                        LockPoints* lp,
-                        bool useExistingLockPoints)
+                             const LockPoints& lp)
```
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/23897#issuecomment-1330960296).

hebasto commented on src/validation.cpp:
```diff
@@ -176,11 +176,67 @@ bool CheckFinalTxAtTip(const CBlockIndex& active_chain_tip, const CTransaction&
+            prevheights[txinIndex] = index.nHeight;
+        } else {
+            prevheights[txinIndex] = coin.nHeight;
+        }
```
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/23897#issuecomment-1330960296).

hebasto commented on src/validation.cpp:
```diff
@@ -176,11 +176,67 @@ bool CheckFinalTxAtTip(const CBlockIndex& active_chain_tip, const CTransaction&
+{
+    assert(tip);
+
+    CBlockIndex index;
```
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/23897#issuecomment-1330960296).

hebasto commented on src/validation.cpp:
```diff
@@ -176,11 +176,67 @@ bool CheckFinalTxAtTip(const CBlockIndex& active_chain_tip, const CTransaction&
+    index.nHeight = tip->nHeight + 1;
+
+    std::vector<int> prevheights;
+    prevheights.resize(tx.vin.size());
```
Going to leave it as this because the suggested diff won't work.

hebasto commented on src/validation.cpp:
```diff
@@ -176,11 +176,79 @@ bool CheckFinalTxAtTip(const CBlockIndex& active_chain_tip, const CTransaction&
+    // *next* block, we need to use one more than active_chainstate.m_chain.Height()
+    next_tip.nHeight = tip->nHeight + 1;
+
+    auto prev_heights = CalculatePrevHeights(next_tip, coins_view, tx);
```
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-auto:

> **Exception** Avoid `auto` for initializer lists and in cases where you know exactly which type you want and where an initializer might require conversion.

stickies-v replied:
Correct me if I'm wrong, but I don't think this applies here. The recommendation you reference is if the type is not specified, e.g. `auto i{1}`. In this case, `CalculatePrevHeights()` has a well defined return type, so list initialization seems appropriate.

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/23897#issuecomment-1410369444).

hebasto commented on src/validation.cpp:
```diff
@@ -176,11 +176,79 @@ bool CheckFinalTxAtTip(const CBlockIndex& active_chain_tip, const CTransaction&
+        const CTxIn& txin = tx.vin[i];
+        Coin coin;
+        if (!coins.GetCoin(txin.prevout, coin)) {
+            LogPrintf("ERROR: %s: Missing input %d in transaction \'%s\'\n", __func__, i, tx.GetHash().GetHex());
+            return std::nullopt;
```
Why? The error message is being logged, and it is not exposed to a user in any other way.

https://github.com/bitcoin/bitcoin/blob/cac29f5cd66f005ce65edd697990d495a904a1f1/src/util/result.h#L22-L26

If I missed your point, what are benefits of propagating the error message back?

stickies-v replied:
Same reason we want to let the callsite decide how to handle what to do when the calculation failed: with `util::Result`, the callsite can also decide if they want to log the error or not. In this case, there is only one callsite so it doesn't matter much. Can always refactor later on if necessary. Happy to mark as resolved, just a thought i had.

hebasto commented:
Updated 1636bc465fd96b9561de3dd21bed7f989f409de1 -> 63d6b9a9f067b81fed6029099a4643996e8d274b ([pr23897.10](https://github.com/hebasto/bitcoin/commits/pr23897.10) -> [pr23897.11](https://github.com/hebasto/bitcoin/commits/pr23897.11)):

- rebased to include #17786.
- addressed more @stickies-v's comments

hebasto commented on src/validation.cpp:
```diff
@@ -176,11 +176,79 @@ bool CheckFinalTxAtTip(const CBlockIndex& active_chain_tip, const CTransaction&
 }
 
+namespace {
+std::optional<std::vector<int>> CalculatePrevHeights(
```
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/23897#issuecomment-1336191549).

hebasto commented on src/validation.cpp:
```diff
@@ -176,11 +176,79 @@ bool CheckFinalTxAtTip(const CBlockIndex& active_chain_tip, const CTransaction&
 
+namespace {
+std::optional<std::vector<int>> CalculatePrevHeights(
+    const CBlockIndex next_tip,
```
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/23897#issuecomment-1336191549).

stickies-v reviewed (approved):
ACK 68c0f16fa2a41d6a25361f0bfa0b74ac5c8c43e2

A couple of improvements for follow-up that I think would further benefit the readability of this code, but I agree with @hebasto's assessment that they may reduce the reviewability of this PR and they are not blocking imo (so comments can be marked resolved, keeping track here):
- [Increase visibility in unexpected error](https://github.com/bitcoin/bitcoin/pull/23897#discussion_r1019009977)
- [carve out `filter_final_and_mature` into separate function](https://github.com/bitcoin/bitcoin/pull/23897#discussion_r1019055149)
- [use `CBlockIndex&` ref instead of pointer](https://github.com/bitcoin/bitcoin/pull/23897#discussion_r1013832021)

stickies-v commented on src/validation.h:
```diff
@@ -245,7 +245,9 @@ PackageMempoolAcceptResult ProcessNewPackage(Chainstate& active_chainstate, CTxM
+ * Calculate LockPoints required to check if transaction will be BIP68 final in the next block
+ * to be created on top of tip.
+ *
  * @param[in]   tip             Chain tip to check tx sequence locks against. For example,
```
E.g. "Chain tip to check tx sequence locks against." -> "Chain tip for which tx sequence locks are calculated."

hebasto replied:
Thanks! [Updated](https://github.com/bitcoin/bitcoin/pull/23897#issuecomment-1410369444).

hebasto committed:
refactor: Add `CalculateLockPointsAtTip()` function

hebasto committed:
refactor: Move calculation logic out from `CheckSequenceLocksAtTip()`

hebasto commented on src/validation.cpp:
```diff
@@ -176,11 +176,67 @@ bool CheckFinalTxAtTip(const CBlockIndex& active_chain_tip, const CTransaction&
+
+    CBlockIndex index;
+    index.pprev = tip;
+    // When SequenceLocks() is called within ConnectBlock(), the height
+    // of the block *being* evaluated is what is used.
```
> I couldn't find any connection between `SequenceLocks()` and `ConnectBlock()` and this function - but maybe I just missed it?

https://github.com/bitcoin/bitcoin/blob/357d750cab5221695d718e6d3e8ce63fa2b5ab3a/src/validation.cpp#L2257

hebasto commented:
Updated 68c0f16fa2a41d6a25361f0bfa0b74ac5c8c43e2 -> 75db62ba4cae048e742ca02dc6a52b3b3d6727de ([pr23897.12](https://github.com/hebasto/bitcoin/commits/pr23897.12) -> [pr23897.13](https://github.com/hebasto/bitcoin/commits/pr23897.13), [diff](https://github.com/hebasto/bitcoin/compare/pr23897.12..pr23897.13)).

Addressed some of @stickies-v's comments:
- https://github.com/bitcoin/bitcoin/pull/23897#discussion_r1038113556
- https://github.com/bitcoin/bitcoin/pull/23897#discussion_r1072450279

stickies-v reviewed (approved):
re-ACK 75db62b

achow101 commented:
ACK 75db62ba4cae048e742ca02dc6a52b3b3d6727de

glozow reviewed (commented):
code LGTM
